batch/sql/add-user-resources-ready-cores-token.sql

```diff
@@ -0,0 +1,171 @@
+ALTER TABLE user_resources ADD COLUMN token INT NOT NULL DEFAULT 0;
+ALTER TABLE user_resources DROP PRIMARY KEY, ADD PRIMARY KEY(`user`, `token`);
+
+ALTER TABLE ready_cores ADD COLUMN token INT NOT NULL DEFAULT 0;
+ALTER TABLE ready_cores ADD PRIMARY KEY(`token`);
+
+CREATE TABLE IF NOT EXISTS `batches_staging` (
+  `batch_id` BIGINT NOT NULL,
+  `token` INT NOT NULL,
+  `n_jobs` INT NOT NULL DEFAULT 0,
+  `n_ready_jobs` INT NOT NULL DEFAULT 0,
+  `ready_cores_mcpu` BIGINT NOT NULL DEFAULT 0,
+  PRIMARY KEY (`batch_id`, `token`),
+  FOREIGN KEY (`batch_id`) REFERENCES batches(`id`) ON DELETE CASCADE
+) ENGINE = InnoDB;
+
+DELIMITER $$
+
+DROP PROCEDURE IF EXISTS insert_batches_staging_tokens;
+CREATE PROCEDURE insert_batches_staging_tokens(
+  IN in_batch_id BIGINT
+)
+BEGIN
+  DECLARE i int DEFAULT 0;
+  DECLARE row_exists BOOLEAN;
+
+  WHILE i < 32 DO
+    SET row_exists = EXISTS(SELECT * FROM batches_staging WHERE batch_id = in_batch_id AND token = i FOR UPDATE);
+    IF NOT row_exists THEN
+      INSERT INTO batches_staging (batch_id, token) VALUES (in_batch_id, i);
+    END IF;
+    SET i = i + 1;
+  END WHILE;
+END $$
+
+DROP PROCEDURE IF EXISTS insert_ready_cores_tokens;
+CREATE PROCEDURE insert_ready_cores_tokens()
+BEGIN
+  DECLARE i int DEFAULT 0;
+  DECLARE row_exists BOOLEAN;
+
+  WHILE i < 32 DO
+    SET row_exists = EXISTS(SELECT * FROM ready_cores WHERE token = i FOR UPDATE);
```

Comment:
With my other suggestion I think this will go away, but for future reference, these tables now have primary keys, so I think you can use REPLACE instead of this SELECT/INSERT.  I think that will be simpler.  (I'm concerned these aren't in an transaction now, and I think this is right but I'd want to double-check the semantics for SELECT ... FOR UPDATE on a non-existent row (it needs to lock the gap)).

Reply:
I used this same sort of code in `add_attempt`. I'll figure out how to check if it takes a gap lock on the row after I address your other comments.

Reply:
I think the code in add_attempt is fine even though it doesn't get a gap lock. If the attempt exists, a row lock will be taken (any row searched in for update is held for the entire transaction even if it's in a subquery -- at least that's my understanding in the docs https://dev.mysql.com/doc/refman/5.7/en/innodb-locking-reads.html). If the attempt doesn't exist, then we don't execute anything anyways so it's ok we don't have the lock.

```mysql
  SET attempt_exists = EXISTS (SELECT * FROM attempts
                               WHERE batch_id = in_batch_id AND
                                 job_id = in_job_id AND attempt_id = in_attempt_id
                               FOR UPDATE);

  IF NOT attempt_exists AND in_attempt_id IS NOT NULL THEN
    INSERT INTO attempts (batch_id, job_id, attempt_id, instance_name) VALUES (in_batch_id, in_job_id, in_attempt_id, in_instance_name);
    SELECT state INTO cur_instance_state FROM instances WHERE name = in_instance_name FOR UPDATE;
    IF cur_instance_state = 'active' THEN
      UPDATE instances SET free_cores_mcpu = free_cores_mcpu - in_cores_mcpu WHERE name = in_instance_name;
      SET delta_cores_mcpu = -1 * in_cores_mcpu;
    END IF;
  END IF;

```